evap/results/views.py

```diff
@@ -62,9 +62,12 @@ def course_detail(request, semester_id, course_id):
     if not request.user.is_staff:
         # remove TextResults if user is neither the evaluated person (or a delegate) nor responsible for the course (or a delegate)
         for section in sections:
-            if not user_can_see_textresults(request.user, course, section):
-                for i, result in list(enumerate(section.results))[::-1]:
-                    if isinstance(result, TextResult):
+            for i, result in list(enumerate(section.results))[::-1]:
+                if isinstance(result, TextResult):
+                    for j, answer in list(enumerate(result.answers))[::-1]:
+                        if not user_can_see_text_answer(request.user, course, answer):
+                            del result.answers[j]
+                    if not result.answers:
```

Comment:
please check whether you can do this easier, maybe in two passes and using the filter builtin or list comprehensions.

something like
```
for section in sections:
    for result in section.results:
        result.answers = [answer for answer in result.answers if user_can_see_text_answer(blah)]
sections = (section for section in sections if section.answers)
```